File: apps/cli/src/commands/cleanup.ts
import { Command } from "commander";

import { formatBold, formatDim, formatError, formatSuccess } from "../lib/colors";
import { isJson, outputError } from "../lib/output";
import { getServerUrl } from "../lib/store";

interface CleanupCandidate {
    id: string;
    title: string;
    type: string;
    reason: string;
    category: "plan-artifact" | "near-empty" | "duplicate-title";
}

function requireServer(): string {
    const serverUrl = getServerUrl();
    if (!serverUrl) {
        console.error('No server URL configured. Run "fubbik init" first.');
        process.exit(1);
    }
    return serverUrl;
}

export const cleanupCommand = new Command("cleanup")
    .description("Identify and optionally remove low-value chunks")
    .option("--dry-run", "show what would be removed (default)", true)
    .option("--confirm", "actually remove flagged chunks")
    .option("--type <type>", "filter to specific chunk type")
    .option("--codebase <name>", "scope to codebase")
    .action(async (opts, cmd) => {
        const serverUrl = requireServer();

        try {
            const params = new URLSearchParams({ limit: "500" });
            if (opts.codebase) params.set("codebaseId", opts.codebase);
            const res = await fetch(`${serverUrl}/api/chunks?${params}`);
            if (!res.ok) {
                outputError(`API error: ${res.statusText}`);
                return;
            }
            const { chunks } = (await res.json()) as { chunks: any[] };

            const candidates: CleanupCandidate[] = [];

            // Filter by type if specified
            const filtered = opts.type
                ? chunks.filter((c: any) => c.type === opts.type)
                : chunks;

            // 1. Plan task artifacts: guide chunks with tags that look like plan filenames
            const planFilePattern = /\d{4}-.*\.md/;
            for (const chunk of filtered) {
                const tags: string[] = Array.isArray(chunk.tags)
                    ? chunk.tags.map((t: any) => (typeof t === "string" ? t : t.name ?? ""))
                    : [];

                if (
                    chunk.type === "guide" &&
                    tags.some((tag: string) => planFilePattern.test(tag) && tag.includes(".md"))
                ) {
                    candidates.push({
                        id: chunk.id,
                        title: chunk.title,
                        type: chunk.type,
                        reason: "Plan task artifact (type: guide, tagged with plan filename)",
                        category: "plan-artifact",
                    });
                    continue; // Don't double-count
                }

                // 2. Near-empty chunks
                if (chunk.content.length < 50) {
                    candidates.push({
                        id: chunk.id,
                        title: chunk.title,
                        type: chunk.type,
                        reason: `Near-empty (${chunk.content.length} chars)`,
                        category: "near-empty",
                    });
                }
            }

            // 3. Duplicate titles
            const titleMap = new Map<string, any[]>();
            for (const chunk of filtered) {
                const key = chunk.title.toLowerCase().trim();
                if (!titleMap.has(key)) titleMap.set(key, []);
                titleMap.get(key)!.push(chunk);
            }
            for (const [, dupes] of titleMap) {
                if (dupes.length > 1) {
                    // Keep the first (oldest), flag the rest
                    const sorted = dupes.sort(
                        (a: any, b: any) =>
                            new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
                    );
                    for (let i = 1; i < sorted.length; i++) {
                        const chunk = sorted[i];
                        // Avoid double-counting if already flagged
                        if (!candidates.some((c) => c.id === chunk.id)) {
                            candidates.push({
                                id: chunk.id,
                                title: chunk.title,
                                type: chunk.type,
                                reason: `Duplicate title (${dupes.length} copies)`,
                                category: "duplicate-title",
                            });
                        }
                    }
                }
            }

            // Count by category
            const planArtifacts = candidates.filter((c) => c.category === "plan-artifact");
            const nearEmpty = candidates.filter((c) => c.category === "near-empty");
            const duplicates = candidates.filter((c) => c.category === "duplicate-title");

            if (isJson(cmd)) {
                console.log(
                    JSON.stringify(
                        {
                            candidates,
                            summary: {
                                planArtifacts: planArtifacts.length,
                                nearEmpty: nearEmpty.length,
                                duplicateTitles: duplicates.length,
                                total: candidates.length,
                            },
                        },
                        null,
                        2
                    )
                );
                if (!opts.confirm) return;
            }

            if (candidates.length === 0) {
                console.error(formatSuccess("No low-value chunks found"));
                return;
            }

            if (!opts.confirm) {
                // Dry-run output
                console.error(formatBold("Cleanup analysis:"));
                if (planArtifacts.length > 0) {
                    console.error(
                        `  ${planArtifacts.length} plan task artifacts (type: guide, tagged with plan filenames)`
                    );
                }
                if (nearEmpty.length > 0) {
                    console.error(
                        `  ${nearEmpty.length} near-empty chunks (< 50 chars)`
                    );
                }
                if (duplicates.length > 0) {
                    console.error(
                        `  ${duplicates.length} duplicate titles`
                    );
                }
                console.error(
                    `\nTotal: ${candidates.length} chunks flagged`
                );
                console.error(
                    formatDim("Run 'fubbik cleanup --confirm' to remove.")
                );
                return;
            }

            // Actually remove
            console.error(formatBold("Removing flagged chunks...\n"));
            let removed = 0;
            let failed = 0;

            for (const candidate of candidates) {
                try {
                    const delRes = await fetch(
                        `${serverUrl}/api/chunks/${candidate.id}`,
                        { method: "DELETE" }
                    );
                    if (delRes.ok) {
                        console.error(
                            formatSuccess(`Removed: ${candidate.title} ${formatDim(`(${candidate.reason})`)}`)
                        );
                        removed++;
                    } else {
                        console.error(
                            formatError(`Failed: ${candidate.title} (${delRes.statusText})`)
                        );
                        failed++;
                    }
                } catch {
                    console.error(
                        formatError(`Failed: ${candidate.title} (network error)`)
                    );
                    failed++;
                }
            }

            console.error(
                `\n${formatSuccess(`${removed} removed`)}${failed > 0 ? ` ${formatError(`${failed} failed`)}` : ""}`
            );
        } catch (e: any) {
            outputError(e.message);
        }
    });

File: crates/fubbik-cli/src/commands/cleanup.rs
use std::collections::{HashMap, HashSet};

use anyhow::{Result, bail};
use serde::Serialize;

use crate::client::Client;
use crate::output::{self, OutputMode};

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
struct Candidate {
    id: String,
    title: String,
    #[serde(rename = "type")]
    chunk_type: String,
    reason: String,
    category: &'static str,
}

pub async fn run(
    client: &Client,
    confirm: bool,
    chunk_type: Option<&str>,
    space: Option<&str>,
    mode: OutputMode,
) -> Result<()> {
    let space_id = client.resolve_space(space).await?;
    let chunks = super::export::load_chunks(client, space_id.as_deref()).await?;
    let candidates = find_candidates(&chunks, chunk_type);
    let result = serde_json::json!({
        "candidates": candidates,
        "summary": {
            "planArtifacts": candidates.iter().filter(|item| item.category == "plan-artifact").count(),
            "nearEmpty": candidates.iter().filter(|item| item.category == "near-empty").count(),
            "duplicateTitles": candidates.iter().filter(|item| item.category == "duplicate-title").count(),
            "total": candidates.len(),
        }
    });
    if !confirm {
        return render_analysis(&result, mode);
    }

    let mut removed = Vec::new();
    let mut errors = Vec::new();
    for candidate in &candidates {
        match client.delete_chunk(&candidate.id).await {
            Ok(_) => removed.push(candidate.id.clone()),
            Err(error) => errors.push(format!("{}: {error}", candidate.title)),
        }
    }
    let applied = serde_json::json!({"removed": removed, "errors": errors});
    if mode == OutputMode::Json {
        output::json(&applied)?;
    } else if mode == OutputMode::Quiet {
        for id in &removed {
            println!("{id}");
        }
    } else {
        println!("{} removed, {} failed", removed.len(), errors.len());
    }
    if !errors.is_empty() {
        bail!("{} chunk(s) could not be removed", errors.len());
    }
    Ok(())
}

fn find_candidates(chunks: &[serde_json::Value], chunk_type: Option<&str>) -> Vec<Candidate> {
    let filtered = chunks
        .iter()
        .filter(|chunk| match chunk_type {
            Some(kind) => chunk["type"] == kind,
            None => true,
        })
        .collect::<Vec<_>>();
    let mut candidates = Vec::new();
    let mut seen = HashSet::new();

    for chunk in &filtered {
        let id = chunk["id"].as_str().unwrap_or("");
        let tags = tag_names(chunk);
        if chunk["type"] == "guide" && tags.iter().any(|tag| is_plan_filename(tag)) {
            candidates.push(candidate(
                chunk,
                "Plan task artifact (type: guide, tagged with plan filename)".into(),
                "plan-artifact",
            ));
            seen.insert(id.to_owned());
        } else if chunk["content"].as_str().map_or(0, str::len) < 50 {
            candidates.push(candidate(
                chunk,
                format!(
                    "Near-empty ({} chars)",
                    chunk["content"].as_str().map_or(0, str::len)
                ),
                "near-empty",
            ));
            seen.insert(id.to_owned());
        }
    }

    let mut titles: HashMap<String, Vec<&serde_json::Value>> = HashMap::new();
    for &chunk in &filtered {
        titles
            .entry(
                chunk["title"]
                    .as_str()
                    .unwrap_or("")
                    .trim()
                    .to_ascii_lowercase(),
            )
            .or_default()
            .push(chunk);
    }
    for duplicates in titles.values_mut().filter(|items| items.len() > 1) {
        duplicates.sort_by_key(|chunk| chunk["createdAt"].as_str().unwrap_or(""));
        let copies = duplicates.len();
        for chunk in duplicates.iter().skip(1) {
            let id = chunk["id"].as_str().unwrap_or("");
            if seen.insert(id.to_owned()) {
                candidates.push(candidate(
                    chunk,
                    format!("Duplicate title ({copies} copies)"),
                    "duplicate-title",
                ));
            }
        }
    }
    candidates
}

fn candidate(chunk: &serde_json::Value, reason: String, category: &'static str) -> Candidate {
    Candidate {
        id: chunk["id"].as_str().unwrap_or("").into(),
        title: chunk["title"].as_str().unwrap_or("Untitled").into(),
        chunk_type: chunk["type"].as_str().unwrap_or("note").into(),
        reason,
        category,
    }
}

fn tag_names(chunk: &serde_json::Value) -> Vec<&str> {
    chunk["tags"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|tag| tag.as_str().or_else(|| tag["name"].as_str()))
        .collect()
}

fn is_plan_filename(tag: &str) -> bool {
    let bytes = tag.as_bytes();
    bytes.len() > 5
        && bytes[..4].iter().all(u8::is_ascii_digit)
        && bytes[4] == b'-'
        && tag.ends_with(".md")
}

fn render_analysis(result: &serde_json::Value, mode: OutputMode) -> Result<()> {
    if mode == OutputMode::Json {
        return output::json(result);
    }
    if result["summary"]["total"] == 0 {
        println!("No low-value chunks found");
        return Ok(());
    }
    if mode == OutputMode::Quiet {
        for item in result["candidates"].as_array().into_iter().flatten() {
            println!("{}", item["id"].as_str().unwrap_or(""));
        }
        return Ok(());
    }
    println!("Cleanup analysis:");
    println!(
        "  {} plan task artifacts",
        result["summary"]["planArtifacts"]
    );
    println!("  {} near-empty chunks", result["summary"]["nearEmpty"]);
    println!(
        "  {} duplicate titles",
        result["summary"]["duplicateTitles"]
    );
    println!("\nTotal: {} chunks flagged", result["summary"]["total"]);
    println!("Run 'fubbik cleanup --confirm' to remove.");
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cleanup_prioritizes_artifacts_and_keeps_oldest_duplicate() {
        // Given a plan artifact and two otherwise substantial chunks sharing a title
        let chunks = vec![
            serde_json::json!({"id":"artifact","title":"Task","type":"guide","content":"","createdAt":"2026-01-01","tags":["2026-01-plan.md"]}),
            serde_json::json!({"id":"old","title":"Shared","type":"note","content":"x".repeat(60),"createdAt":"2026-01-01","tags":[]}),
            serde_json::json!({"id":"new","title":" shared ","type":"note","content":"x".repeat(60),"createdAt":"2026-02-01","tags":[]}),
        ];
        // When cleanup candidates are classified
        let candidates = find_candidates(&chunks, None);
        // Then the artifact is not double-counted as empty and only the newer duplicate is removed
        assert_eq!(candidates.len(), 2);
        assert!(
            candidates
                .iter()
                .any(|item| item.id == "artifact" && item.category == "plan-artifact")
        );
        assert!(
            candidates
                .iter()
                .any(|item| item.id == "new" && item.category == "duplicate-title")
        );
    }
}
